extensions/interactions/MusicNotesInput/directives/oppia-interactive-music-notes-input.component.ts

```diff
@@ -0,0 +1,871 @@
+        'MusicNotesInput', this._getAttributes()
+      ) as MusicNotesInputCustomizationArgs);
+
+    this.sequenceToGuess = sequenceToGuess;
+    this.interactionIsActive = (this.lastAnswer === null);
+
+    this.initialSequence = this.interactionIsActive ?
+      initialSequence : this.lastAnswer;
+
+    this.directiveSubscriptions.add(
+      this.playerPositionService.onNewCardAvailable.subscribe(() => {
+        this.interactionIsActive = false;
+        this.initialSequence = this.lastAnswer;
+        this.reinitStaff();
+      })
+    );
+
+    this.currentInteractionService.registerCurrentInteraction(
+      () => this.submitAnswer(), null);
+    // Initialization code.
+    this.initializeNoteSequence(this.initialSequence as Sequence);
+    this.init();
+  }
+
+  ngAfterViewInit(): void {
+    // Sets grid positions, displays the staff and note,
+    // and then initializes the view after staff has loaded.
+    this.reinitStaff();
+  }
+
+  /**
+   * A note Object has a baseNoteMidiNumber and an offset property. For
+   * example, C#4 would be -> note = {baseNoteMidiNumber: 61, offset: 1};
+   *
+   * A readableNote Object has a readableNoteName property. For example,
+   * Gb5 would be -> readableNote = {readableNoteName: 'Gb5'};
+   */
+
+  /**
+   * Array to hold the notes placed on staff. Notes are represented as
+   * objects with two keys: baseNoteMidiNumber and offset. The
+   * baseNoteMidiNumber is an integer value denoting the MIDI number of
+   * the staff-line the note is on, and the offset is either -1, 0 or 1,
+   * denoting a flat, natural or sharp respectively.
+   */
+  _addNoteToNoteSequence(note: MusicNote): void {
+    this.noteSequence.push({
+      note: note
+    });
+  }
+
+  // Remove a specific note with given noteId from noteSequence. If given
+  // noteId is not in noteSequence, nothing will be removed.
+  _removeNotesFromNoteSequenceWithId(noteId: string): void {
+    for (let i = 0; i < this.noteSequence.length; i++) {
+      if (this.noteSequence[i].note.noteId === noteId) {
+        this.noteSequence.splice(i, 1);
+      }
+    }
+  }
+
+  // Sorts noteSequence elements according to the return value of the
+  // compareNoteStarts function.
+  _sortNoteSequence(): void {
+    this.noteSequence.sort(this.compareNoteStarts);
+  }
+
+  /**
+   * Counter to create unique id for each note dropped on staff.
+   */
+  generateNoteId(): string {
+    this._currentNoteId += 1;
+    return 'note_id_' + (this._currentNoteId - 1);
+  }
+
+  // Staff has to be reinitialized every time that the staff is resized or
+  // displayed. The staffContainerElt and all subsequent measurements
+  // must be recalculated in order for the grid to work properly.
+  reinitStaff(): void {
+    $('.oppia-music-input-valid-note-area').css('visibility', 'hidden');
+    setTimeout(() => {
+      $('.oppia-music-input-valid-note-area').css(
+        'visibility', 'visible');
```

Comment:
We should reduce the usage of jQuery can you add TODO here?

Reply:
ditto elsewhere in this file

Reply:
done